batch2/batch/driver/scheduler.py

```diff
@@ -38,37 +39,87 @@ def __init__(self, app):
                 await changed.wait()
                 changed.clear()
 
-    async def cancel_1(self):
+    async def get_n_jobs_per_user(self, max_n_jobs):
         records = self.db.execute_and_fetchall(
             '''
+SELECT user, ready_n_jobs, ready_cores_mcpu
+FROM users;
+''')
+
+        ready_users = {record['user']: record['ready_n_jobs'] for record in records}
+
+        result = {user: 0 for user in ready_users}
+        n_jobs_to_allocate = max_n_jobs
+
+        if 'ci' in ready_users and n_jobs_to_allocate > 0:
+            n_ci_jobs = min(ready_users['ci'], n_jobs_to_allocate)
+            result['ci'] = n_ci_jobs
+            n_jobs_to_allocate -= n_ci_jobs
+            del ready_users['ci']
+
+        if 'test' in ready_users and n_jobs_to_allocate > 0:
+            n_test_jobs = min(ready_users['test'], n_jobs_to_allocate)
+            result['test'] = n_test_jobs
+            n_jobs_to_allocate -= n_test_jobs
+            del ready_users['test']
+
+        other_users = random.shuffle(list(ready_users.keys()))
+
+        while n_jobs_to_allocate > 0 and len(other_users) > 0:
+            k = len(other_users)
+            n = n_jobs_to_allocate
+            jobs_per_user = [(n // k) + (1 if i < (n % k) else 0) for i in range(k)]
+
+            for i, user in enumerate(other_users):
+                n_user_jobs = min(jobs_per_user[i], ready_users[user])
+                result[user] += n_user_jobs
+                ready_users[user] -= n_user_jobs
+                if ready_users[user] == 0:
+                    del ready_users
+
+            other_users = random.shuffle(list(ready_users.keys()))
+
+        return result
+
+    async def get_records(self, f, max_jobs):
+        jobs_per_user = await self.get_n_jobs_per_user(max_jobs)
```

Comment:
I think you can get rid of `user_n_jobs` and just query `LIMIT 50` or whatever, and if you get fewer jobs, you get fewer jobs.